File: Assets/Scripts/NativeTest/NativeTest.cs
using UnityEngine;

namespace Cicanci.NativeTest
{
    public class NativeTest
    {
        private readonly INativeTest _nativeTest;

        public NativeTest()
        {
#if UNITY_EDITOR
            _nativeTest = new NativeTestDummy();
#elif UNITY_ANDROID
            _nativeTest = new NativeTestAndroid();
#endif
        }

        public int GetCounter()
        {
            return _nativeTest.GetCounter();
        }

        public string GetMessage()
        {
            return _nativeTest.GetMessage();
        }

        public void SetCounter(int counter)
        {
            _nativeTest.SetCounter(counter);
        }

        public void SetMessage(string message)
        {
            _nativeTest.SetMessage(message);
        }
    }
}

File: Assets/Plugins/Android/native-test-lib/src/com/cicanci/unity/NativeTest.java
package com.cicanci.unity;

public class NativeTest {

    private int counter = 0;
    private String message = "Hello from Android";

    public int getCounter() {
      return ++this.counter;
    }

    public String getMessage() {
      return this.message;
    }

    public void setCounter(int counter) {
      this.counter = counter;
    }

    public void setMessage(String message) {
      this.message = message;
    }

}
